pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+	instData.ipaclattr = make(map[int]map[int]aclAttr) // App-ID/ACL-Num/aclAttr table
+	instData.appIPAddrs = make(map[int][]net.IP)
+	instData.bnNet = make(map[string]aclAttr) // borrow the aclAttr for intf attributes
+	instData.appNet = make(map[int]uuid.UUID)
+	pub := ctx.pubAppFlowMonitor
+
+	currentEntries := ctx.subNetworkInstanceConfig.GetAll()
+	for key, entry := range currentEntries {
+		config := cast.CastNetworkInstanceConfig(entry)
+		log.Infof("FlowStatsCollect: on NetworkInstance %s (Name: %s) \n",
+			key, config.DisplayName)
+	}
+
+	IntfAddrs, err := net.InterfaceAddrs()
+	if err != nil {
+		log.Errorf("error in getting addresses\n")
+		return
+	}
+	instData.intfAddrs = IntfAddrs
+
+	if devUUID == nilUUID {
+		b, err := ioutil.ReadFile("/config/uuid")
+		if err != nil {
+			log.Errorf("error in reading uuid\n")
+			return
+		}
+		uuidStr := strings.TrimSpace(string(b))
+		devUUID, err = uuid.FromString(uuidStr)
+		if err != nil {
+			log.Errorf("error in formating uuid\n")
+			return
+		}
+	}
+
+	checkAppAndACL(ctx, instData)
+
+	idx := 0
+	for appNum := range instData.ipaclattr {
+		for aclNum, val := range instData.ipaclattr[appNum] {
+			log.Infof("@@@FlowStats:ACL postcheck(%d) AppNum %d, aclNum %d, table %s, acl-name %s, action %s",
+				idx, appNum, aclNum, val.tableName, val.aclName, val.action)
+			idx++
+		}
+	}
+	//var Protocols [2]netlink.InetFamily
+	//Protocols[0] = syscall.AF_INET
+	//Protocols[1] = syscall.AF_INET6
+
+	//var flowtmp flowStats // XXX debug
+	var timeOutTuples []flowStats
+	var totalFlow int
+	var dnsPacked bool
+	Protocols := [2]netlink.InetFamily{syscall.AF_INET, syscall.AF_INET6}
+
+	// Get IPv4/v6 conntrack table flows
+	for _, proto := range Protocols {
+		connT, err := netlink.ConntrackTableList(netlink.ConntrackTable, proto)
+		if err != nil {
+			log.Errorf("FlowStats(%d): ContrackTableList", proto)
+			return
+		}
+
+		log.Infof("***FlowStats(%d): device=%v, size of the flows %d\n", proto, devUUID, len(connT))
+
+		for _, entry := range connT { // loop through and process current timedout flow collection
+
+			flowTuple := flowMergeProcess(entry, instData)
+			// flowTuple := FlowMergeTuple(entry, instData, ipToName)
+			if flowTuple.IsTimeOut == false {
+				//log.Infof("FlowStats ++stay++ [%d]: %s\n", i, entry.String()) // XXX remove
+				continue
+			}
+
+			timeOutTuples = append(timeOutTuples, flowTuple)
+			totalFlow++
+		}
+	}
+
+	log.Infof("FlowStats ++ Total timedout flows %d, loopcount debug %d\n", totalFlow, loopcount)
+	loopcount++
+
+	// per app/bridge packing flow stats to be uploaded
+	for bnx := range instData.bnNet {
+		log.Infof("FlowStats: loop for bnx %s\n", bnx)
+		for appIdx := range instData.appNet {
+
+			scope := types.FlowScope{
+				UUID:      instData.appNet[appIdx],
+				Intf:      bnx,
+				Localintf: instData.bnNet[bnx].vif,
+				NetUUID:   instData.bnNet[bnx].netUUID,
```

Comment:
Should we check if app is even connected to this bridge (network instance) first?

Reply:
here i just fill in the data first, the real check comes later, if ipaclattr, given app/bridge we can not find the aclnum, then we'll continue the loop without process/upload this data filled here.